backend/guides/grpc.md

```diff
@@ -0,0 +1,158 @@
+# ↔️ gRPC styleguide
+
+## 📐 Design conventions
+
+- Always use custom request / response object even if it should be empty (the only exception is `Delete`, which returns `google.protobuf.Empty`) [📑](https://cloud.google.com/apis/design/design_patterns#empty_responses)
+- In each protofile define package and java package [📑](https://cloud.google.com/apis/design/naming_convention#package_names), e.g.
+
+```proto
+package ackee.newapp.v1;
+option java_package = "cz.ackee.newapp.v1";
+```
+
+- Prefer unary messages with repeated messages to streams (implementation of flow control is not mature enough)
+
+### Metadata
+
+Since there is no gRPC standard for metadata, treat it as HTTP headers. Analogy is accurate since both client and server can send receive metadata with each call.
+
+Use metadata for data that can be generally applicable to all (most) calls - authentication, language, tracking information etc.
+
+When using server-sent metadata, prefer initial to trailing. Since we use only unary calls, we don't need the distinction, and we faced issues with trailing metadata in combination with Envoy proxy.
+
+
+### Different shapes of message
+
+If you want to create few types of the same message (e.g. bare on list, rich on detail), don't create new messages! [📑](https://cloud.google.com/apis/design/naming_convention#message_names) Define the message as its largest subset of fields and go for either of...
+
+1. Create a loose contract about sending this and that (verbally, documented)
```

Comment:
Is it possible to present the example with reuse in list and detail? I have no idea what is meant by either of the two options :(

Reply:
> Define the message as its largest subset of fields

We assume the two shapes of the same thing don't have conflict on fields. (e.g. `A` would have string `id` and `A'` would have a composite object on field `id`). In that case you can create a message `M = A ⋂ A'`. Use for both list and detail `M`. All attributes are required, but have default value if you don't set them.

1. Tell clients "on list, dont touch article body, its empty but just because we don't set it, you need to fetch detail for that"
2. Tell clients "we fill list and detail message with different data, to tell which is which and what has what, see the resource views"